### PR TITLE
Add JS-specific diagnostic message for `resolve()` in Promise where type argument can't be inferred

### PR DESCRIPTION
**Pull request by nicdard**
Fixes: https://github.com/microsoft/TypeScript/issues/46570

Continues https://github.com/microsoft/TypeScript/pull/46637
I have cherry picked the commit from there.
Address the comments of the above PR, restricting the change to JS files.

### CODE REVIEW

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -3365,6 +3365,10 @@
         "category": "Error",
         "code": 2809
     },
+    "Expected {0} arguments, but got {1}. TypeScript may need a JSDoc hint that the call to 'new Promise()' produces a 'Promise<void>'": {
```
Comment (jakebailey) [inline src/compiler/diagnosticMessages.json:3368]: I believe this was the message [suggested in the issue](https://github.com/microsoft/TypeScript/issues/46570#issuecomment-955106669), so it make sense to use it, but this wording strikes me as off as I'm not sure that we have any error messages that effectively "blame" TypeScript itself for needing a hint about the code's behavior, at least not in this particular wording. I would have written something more like

> This promise may need a JSDoc-style type that the call to 'new Promise()' produces a 'Promise<void>'

To shift things around to pointing at the particular code that's at fault, not at the language itself.

But, this is bikeshedding. @DanielRosenwasser @andrewbranch

**Review by andrewbranch — CHANGES_REQUESTED** [review]

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -30266,8 +30266,11 @@ namespace ts {
             const parameterRange = hasRestParameter ? min
                 : min < max ? min + "-" + max
                 : min;
+            const file = getSourceFileOfNode(node);
             const error = hasRestParameter ? Diagnostics.Expected_at_least_0_arguments_but_got_1
-                : parameterRange === 1 && args.length === 0 && isPromiseResolveArityError(node) ? Diagnostics.Expected_0_arguments_but_got_1_Did_you_forget_to_include_void_in_your_type_argument_to_Promise
+                : parameterRange === 1 && args.length === 0 && isPromiseResolveArityError(node) ? (file.scriptKind === ScriptKind.JS || file.scriptKind === ScriptKind.JSX)
```
[the remark this answers is not in the document]
Comment (nicdard) [reply src/compiler/checker.ts:30271]: Ok thanks!

File: src/compiler/diagnosticMessages.json (hunk 1 above)
Comment (nicdard) [reply src/compiler/diagnosticMessages.json:3372]: It adds a type tag: `/** @type {Promise<void>} */(new Promise(resolve => resolve()));`

Comment (andrewbranch) [reply src/compiler/diagnosticMessages.json:3372]: How about

```
'new Promise()' needs a JSDoc hint to produce a 'resolve' that can be called without arguments.
```

- `'resolve'` could be substituted with an interpolation to use the parameter name the user wrote.
- The first interpolation (`{0} arguments`) is unnecessary as we know it’s `0`.

(cc @DanielRosenwasser)

**Comment by andrewbranch** [comment]
In a separate PR, we should consider expanding the codefix to cover

```ts
/** @param {() => void} resolve */
function alwaysResolve(resolve) { resolve() }
new Promise(alwaysResolve);
```

(both JS and TS versions)

**Comment by nicdard** [comment]
> In a separate PR, we should consider expanding the codefix to cover
> 
> ```ts
> /** @param {() => void} resolve */
> function alwaysResolve(resolve) { resolve() }
> new Promise(alwaysResolve);
> ```
> 
> (both JS and TS versions)

Doesn't TS complain about type of resolve function (which should take an argument) not macthing the type of alwaysResolve?

**Comment by andrewbranch** [comment]
Yes. It’s a different error but the same fix can be applied.

File: src/compiler/diagnosticMessages.json (hunk 1 above)
Comment (nicdard) [reply src/compiler/diagnosticMessages.json:3372]: Nice, this provides better feedback indeed :)

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -30413,11 +30413,15 @@ namespace ts {
...
                 : min < max ? min + "-" + max
                 : min;
-            const error = hasRestParameter ? Diagnostics.Expected_at_least_0_arguments_but_got_1
-                : parameterRange === 1 && args.length === 0 && isPromiseResolveArityError(node) ? isInJSFile(node)
-                    ? Diagnostics.Expected_0_arguments_but_got_1_TypeScript_may_need_a_JSDoc_hint_that_the_call_to_new_Promise_produces_a_Promise_void
-                    : Diagnostics.Expected_0_arguments_but_got_1_Did_you_forget_to_include_void_in_your_type_argument_to_Promise
-                : Diagnostics.Expected_0_arguments_but_got_1;
+            const isVoidPromiseError = !hasRestParameter && parameterRange === 1 && args.length === 0 && isPromiseResolveArityError(node);
+            if (isVoidPromiseError && isInJSFile(node)) {
+                return getDiagnosticForCallNode(node, Diagnostics.Expected_1_argument_but_got_0_new_Promise_need_a_JSDoc_hint_to_produce_a_0_that_can_be_called_without_arguments, "resolve");
```
Comment (nicdard) [inline src/compiler/checker.ts:30418]: Is there any convenient way to get the name of the function? I tried `symbolName(node.symbol)` without success

Comment (andrewbranch) [reply src/compiler/checker.ts:30418]: `resolveName(node.expression, node.expression.escapedText, SymbolFlags.Value, undefined, undefined, false)?.name`

File: src/compiler/diagnosticMessages.json [hunk 4]
```diff
@@ -3369,6 +3369,10 @@
         "category": "Error",
         "code": 2809
     },
+    "Expected 1 argument, but got 0. 'new Promise()' need a JSDoc hint to produce a {0} that can be called without arguments.": {
```
Comment (andrewbranch) [inline src/compiler/diagnosticMessages.json:3372]:
```suggestion
    "Expected 1 argument, but got 0. 'new Promise()' needs a JSDoc hint to produce a {0} that can be called without arguments.": {
```

File: src/compiler/checker.ts (hunk 3 above)
Comment (andrewbranch) [reply src/compiler/checker.ts:30418]: It may also be reasonable just to call it `resolve` since that’s the parameter name in the library file.

Comment (nicdard) [reply src/compiler/checker.ts:30418]: Ok, I will change it to a straightforward string

File: src/compiler/diagnosticMessages.json (hunk 4 above)
Comment (nicdard) [reply src/compiler/diagnosticMessages.json:3372]: Oh sorry for this!

**Review by andrewbranch — APPROVED** [review]
Thank you!

**Pull request merged**